Save to Journal ++

srevinsaju opened the pull request:
@JuiP The commits might be bad; I am attempting to fix something. ~The Pr might not work atm, But I will let you know when its ready~. Its ready :tada:

=========

What works 
* Saving and recalling from Journal 
If you are testing it, please right click on activity, Click `Start New`; Because the old journal data might clash with the current working, which does not give out correct test results.

What is to be fixed (I might not do that in this PR, maybe later):
* When the train is restored, another train is created. This is possibly because, on relaunch, the Bridge activity does not look for any more trains on the current display. I hope that would be easy to fix 

* On restore, the stress goes awry (to maybe 1000%), it might be possible to not read stress, or restart from 0 only for stress

* ~Cost is not saved (maybe I can add a fix for that too, later on another PR)~. Done on https://github.com/JuiP/Bridge/pull/2/commits/0940001e0da03a41df211d075b2f73e05f5792d7

* On restore, there are green, but smaller circles on the wheels, which I am not sure why it happens.

srevinsaju committed:
Merge physics/myelements to bridge/myelements

srevinsaju committed:
Add sugar cursor

srevinsaju committed:
Provide the self activity as args

srevinsaju committed:
Handle sugar cursor signal callback by setting canvas to game

srevinsaju committed:
Some debug imports

srevinsaju committed:
Merge necessary changess from physics/elements/add_object

srevinsaju committed:
Minor formatting

srevinsaju committed:
Rename input to input_unit (upstream/physics)

srevinsaju committed:
get_color -> private function (upstream/physics)

srevinsaju committed:
Upstream physics changes

srevinsaju committed:
Rewrite read and write world data parser function

srevinsaju committed:
Add a conditional to serialize json output if asked for

srevinsaju committed:
Add some imports

srevinsaju committed:
Select appropriate lines of physics (upstream) __init__

Copied to Bridge/physics __init__

srevinsaju committed:
Remove pygame.init, as it is handled by sugargame

srevinsaju committed:
Add option to show sugar cursor within pygame renderer (physics)

srevinsaju committed:
Move write_file and read_file

srevinsaju committed:
Add a newline (formatting)

srevinsaju committed:
Rewrite the opening queue function to parse written data shapes

srevinsaju committed:
Do not start clock on run(), start at __init__()

srevinsaju committed:
Add callback functions to change cursor for UI consistency

srevinsaju committed:
Re-add the read_file() and write_file() methods to match the position on physics (upstream)

srevinsaju committed:
Initialize Pygame font, if the pygame display init gets delayed

srevinsaju committed:
Remove redundant / outdated comment

srevinsaju committed:
Add method to set box2d fps (physics/upstream)

srevinsaju committed:
Reorder joystick object definition

srevinsaju committed:
Reorder toollist initialization

srevinsaju committed:
Demote debugging to False, also rely on env vars for debug bool change

srevinsaju committed:
Connect sugar cursor to signal for UI Consistency

srevinsaju committed:
Add support for saving cost to journal

JuiP commented:
Hi, Thanks a lot for helping :)
Tested,

- The cost increases by 10 on just clicking randomly anywhere
- After using the joints using sugar cursor, the cursor disappears for a long time and does not reappear.
- I got an error with the sounds.
- The cursor movement is really slow.

I haven't reviewed the changes you made yet. Will do that soon.
Also do tell me what you plan to do, do you want to make more changes or want me to use your commits and complete my PR [#26](https://github.com/sugarlabs/Bridge/pull/26)? If you are willing to complete the saving to Journal part, then I suggest you make a pull request to [sugarlabs](https://github.com/sugarlabs/Bridge) that way, everyone who wishes to review can help :)

@chimosky @quozl Please review :)

srevinsaju commented:
@JuiP 

> The cost increases by 10 on just clicking randomly anywhere

That is because, on each click, a circle is created. As each circle is a resource, and hence needs to be paid for

> After using the joints using sugar cursor, the cursor disappears for a long time and does not reappear.

Need to see that

> I got an error with the sounds.

No clue;

> The cursor movement is really slow.

It depends on `pygame.clock.tick`. Check if the cursor movement is slow on `physics` too.

> I haven't reviewed the changes you made yet. Will do that soon.

In case you find any errors regarding to formatting, please necessarily cherry pick and fix formatting; I might not spend extra time on this PR

> Also do tell me what you plan to do, do you want to make more changes or want me to use your commits and complete my PR #26? If you are willing to complete the saving to Journal part, then I suggest you make a pull request to sugarlabs that way, everyone who wishes to review can help :)

This will possibly create a merge conflict when one of our PR is merged. I would not have created this PR, if https://github.com/sugarlabs/Bridge/pull/26#issuecomment-629622605 and https://github.com/sugarlabs/Bridge/pull/26#issuecomment-629622605 were not created. I only created this fix so that this PR can be completed (soon enough). Also, as the possibility of a nearby release with the PR has diminished, I assumed the team would not merge this PR without the refresh from Journal feature.

>  If you are willing to complete the saving to Journal part, then I suggest you make a pull request to sugarlabs that way, everyone who wishes to review can help

You can probably see the difference of this PR from the rest of my PRs on sugarlabs which are neatly annotated and somewhat concerned about flake8 formatting. However this PR is a collection of aggregate changes on my desktop during GCI which I could not complete perfectly. In case that would help. I might not check the reviews on this PR even if it happens. Any PR on sugarlabs/repo should be done with quality, concerning commit annotation, whitespace and code formatting, and finally, the PR should work!. As this PR doesn't work satisfactorily, there is no point of asking for a review on `sugarlabs/Bridge` or asking for a merge when its not working satisfactorily, and I just wanted to make sure it helps 


Please select the best parts of the PR. In the case I know how to fix the error for sure, I would have created the PR on sugarlabs/Bridge before you did during December / January. As Bridge required some extra hours of time; I couldn't spend more time on it. 

Also, my general note, in the case you do not want this PR, feel free to close it. This is not some extra work I put in now, but rather a collection of my unfinished GCI work. When rebased on top of your PR https://github.com/sugarlabs/Bridge/pull/26, it worked somewhat.

srevinsaju commented:
> The cost increases by 10 on just clicking randomly anywhere

I can reproduce on your `Fix` branch too. 
Is it a bug or a feature?

srevinsaju committed:
Reorder imports

srevinsaju committed:
Demote stdout to logging.info

srevinsaju committed:
Remove unwanted imports (#17)

Some more flake8 errors which include unused imports
which remained unfixed on #17

srevinsaju committed:
Fix cursor redraw on every pygame.event

srevinsaju commented:
Fixed the cursor bug on https://github.com/JuiP/Bridge/pull/2/commits/265d6a7537f0aeb1dccd9e1bbc6c22e90f35ea8b

quozl commented:
I've looked at the discussion, but not the commits.  I understand this was unpublished work from a while ago, with some recent change.  @JuiP, you can choose for yourself whether any of these commits can and should be applied to your branch using cherry-pick or hand editing.  You would do that by reviewing each commit; not by trying to apply it ... unless you understand it.  @srevinsaju, you could also do that, but it gets complicated and could be time consuming.  Not saying you shouldn't, but @JuiP has the open pull request on Bridge at the moment, so I'd be inclined to take timing advice from her.  A big merge is messy for reading history and future use of `git bisect`.

JuiP commented:
Sure! I haven't got the time to extensively review and test the commits yet. I was occupied - working on other activities, I plan on getting back to Bridge Activity in a day or two! :)

srevinsaju commented:
> @srevinsaju  you could also do that, but it gets complicated and could be time consuming

Yes @quozl  I will do that in case @JuiP 's  PR gets merged to master in the first place. Because resolving merge conflicts are the hardest things I have done on git, and I have felt this is better;  in case the files diverge very heavily.

quozl commented:
Yes, it is hard.  Great skill to learn though.  If I may advise you;
* try it when you don't need to do it, so you can learn without social bias,
* try it in a recursive copy of your repository, such as in /tmp, made using `cp -ar`; many people who try it the first few times in their own repository end up making a mess they don't know how to fix,
* when reviewing conflict markers inside a file, display the patch you are trying to merge, and understand what caused the conflict,
* rather than merge, look for other ways to solve, like cherry-picking from branch A to branch B, or vice versa.

JuiP commented:
I reviewed and tested these changes;
I have picked the commits which seem correct to me and applied them on branch Fix. Thanks @srevinsaju! :smile: 

While testing your branch I observed:

- [ ] The joints, after turning green are positioned somewhere else.

- [ ] Stress seems to increase randomly and in very large numbers, when resumed from journal

- [ ] The cost is not added after resuming from journal. I changed your implementation of saving cost in `Fix`.

- [ ] I get an error for playing sounds.
